### ISSUE TITLE
Distributed

### ISSUE THREAD

**Issue by lucaswiser** [datamade/dedupe#427]
Anyone working on making this library work in a distributed environment? Seems like it would be great for a Spark library. Basically use blocking to assign partitions

**Comment by kshishir** [comment]
I agree. It would be great to be able to run it in Spark.

**Comment by fgregg** [comment]
Sounds cool. This is not something I personally will be working on, but I'd be happy to support an effort led by someone else.

**Comment by lucaswiser** [comment]
I'm thinking about working on it, though I am really a spark beginner. If someone knows more about it I would love to hear their opinions. A very naive idea is to just do an initial blocking in spark, which would assign a group to a slave node, where the standard dedupe package would run (and do its standard blocking on each cluster to further remove # comparisons, and then combine the results via some reduce operation to send it back to the driver operation. Does anyone know if that would work? Biggest issue with that design is the loss of recall from the initial blocking, but you could potentially have some information sharing between clusters which would allow you to do overlapping blocks. Advantage here is that you wouldn't have to rewrite the entire package in scala.

On a side note, what are standard run-times that people are experiencing using the standard dedupe package? Could a 1m X 1m record linkage run be done on a single laptop (2-core, 4 cpu, 16 gb RAM) in a reasonable amount of time (say an hour)? I am finding my programs are running very slow. I also tried to use the CRF edit distance on a tiny data set (1000 data points) and after 30 minutes I gave up and had to kill the program.

**Comment by fgregg** [comment]
That's pretty slow.

On Wed, Nov 25, 2015 at 11:34 AM, lucaswiser <notifications@github.com>
wrote:

> I'm thinking about working on it, though I am really a spark beginner. If
> someone knows more about it I would love to hear their opinions. A very
> naive idea is to just do an initial blocking in spark, which would assign a
> group to a slave node, where the standard dedupe package would run (and do
> its standard blocking on each cluster to further remove # comparisons, and
> then combine the results via some reduce operation to send it back to the
> driver operation. Does anyone know if that would work? Biggest issue with
> that design is the loss of recall from the initial blocking, but you could
> potentially have some information sharing between clusters which would
> allow you to do overlapping blocks. Advantage here is that you wouldn't
> have to rewrite the entire package in scala.
>
> On a side note, what are standard run-times that people are experiencing
> using the standard dedupe package? Could a 1m X 1m record linkage run be
> done on a single laptop (2-core, 4 cpu, 16 gb RAM) in a reasonable amount
> of time (say an hour)? I am finding my programs are running very slow. I
> also tried to use the CRF edit distance on a tiny data set (1000 data
> points) and after 30 minutes I gave up and had to kill the program.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/datamade/dedupe/issues/427#issuecomment-159682927>.
>



-- 
773.888.2718

**Comment by fgregg** [comment]
CRF distance is much slower than the default edit distance, that's why it's not the default option, but I'm surprised it's so slow for you. I would expect that you should be able to do both tasks on a machine with those specs.

**Comment by lucaswiser** [comment]
Ok so not using CRF, the 1mil X 1mil data set should be feasible within an hour on my laptop? Also, i am really having trouble getting the multiprocessing on os-x to work. I've followed the instructions in the docs, and rebuilt numpy with openblas, but even after reinstalling dedupe it still thinks numpy is linked against the accelerate framework. I've tried this in virtualenv's (using conda) but still no luck. Maybe conda is the issue? Sorry for posting this off topic! 

In any case I have been researching running dedupe in spark, and it does look like you can create custom partitions and run functions confined to each partition. Might have to do some of it in scala however.

**Comment by kshishir** [comment]
I have run about 500k on just the name field. 
On a Linux machine with 16cores and 32GB memory. It runs in about couple of hours. 
 I need good match on names so I have left the CRF on for Name and String. 
I am still evaluating and trying to understand different Dedupe options to use for my data. 

I am also learning pySpark on the side. I have a machine with pySpark on Hadoop clusters. 
After I understand the dedupe process more and get myself familiar with spark I am planning to work on making dedupe run on spark.

I will get in touch with you more in a week or so. 

-Shishir

**Comment by fgregg** [comment]
it looks like it is linked against accelerate

```
extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
```

**Comment by lucaswiser** [comment]
Right but I followed the instructions in the docs, and as you can see am using the openblas library, so I don't understand why it is still linked. Is there some issue with having installed numpy with conda, then uninstalled using pip and reinstalled from source?

**Comment by fgregg** [comment]
dunno. @evz thoughts?

**Comment by lucaswiser** [comment]
Ok so I have taken a first stab at using dedupe with spark and have some good initial results. Basically the naive idea of implementing this is to try and use as much of python dedupe as possible and spark as little as possible to get this to work. This will allow us to get a semi workable solution and see where to go if we want to dive into the Spark code and flesh this out. 

Here is the basic idea: we use Spark to split up the dataset into partitions, and then use vanilla dedupe on each partition (you could have 1 partition per core, or 1 partition per machine and let dedupe use multiprocessing). The key idea here is we use **blocking** to assign partitions. That way we can hope that most of the duplicates are contained on each partition, in which case we can use standard dedupe without **any** network data transfers. How good we can do partitioning will decide the how much recall we lose. I have tested this on a toy dataset, and it looks pretty promising. 

I'll include the files below. I basically took the record_linkage_example from the dedupe docs and wrapped the basic functionality in a def main(), which takes in a pandas dataframe containing two datasets we want to record link against (there is a 'group' column in the df that says which dataset each record actually belongs to). I then imported this function into an ipython notebook, and used this function to compute the record linkage on 2 partitions I created with spark. I partitioned them so that all of the duplicates (which I knew from ID's) were on the same partition. 

Main results: 
It worked! While really simple I think this shows the promise of this idea. The results returned by spark were identical to the ones returned by running the dedupe program on each grouping separately. Furthermore, using Spark **all of my cores were used** which as I've mentioned on this thread I can't do in standard dedupe because I'm on a mac. The standard run took 51 seconds while the spark run took 35. The great thing about spark is that the code you write to test on a local machine is mostly identical to what you would use on a cluster. So this example could be used in a distributed environment.

Things to work on:
Input/blocking:
In record linkage we need to be able to partition both data sets so that the similar items from each end up on the same block. I got around this by simply putting both datasets together (in a pandas df with a group flag) and using a separate dataframe for each partition. Obviously we would want to improve this, and we would want to implement actual blocking. I'm not really if spark can operate on multiple RDD's in the way we want so we might have to stick with combining the data, and leaving a group flag to later split the data set back into to separate sets.

Training:
This is another big one. In my example I used some already prepared training data that I had stored in a file where I am running the spark programs. **Both** spark processes had access to this file, which they used for training. In general this wouldn't be the case because the processes will be on different machines. You could possibly just put a copy of the training data on each machine, but really we want the active learning portion of dedupe to work. This I have no idea about. I'm not sure spark processes can interact with the driver programs terminal like standard dedupe to ask the user to clarify matches. Maybe you could start a server on each remote machine which would generate matches, and then you could visit each of the servers using a browser and clarify matches. 

Here are the two files. They are html files displaying ipython code (they are in .txt so I could upload them. If you want to look at them download and change extension to .html then open in a browser). The dedupe_wrap shows a wrapped record linkage script and the spark_dedupe shows the ipython notebook working through the example. 

[dedupe_wrap.txt](https://github.com/datamade/dedupe/files/46259/dedupe_wrap.txt)
[spark_dedupe.txt](https://github.com/datamade/dedupe/files/46260/spark_dedupe.txt)

**Comment by fgregg** [comment]
What version of python are you on?

**Comment by fgregg** [comment]
cool. I'll look at this more.

**Comment by fgregg** [comment]
@lucaswiser for your accelerate issue try, setting


```extra_link_args = []```

in the numpy site.cfg

**Comment by kshishir** [comment]
Great! I'll give it a try. 

-Shishir

**Comment by lucaswiser** [comment]
I'm using python 2.7. Tried adding the line 
extra_link_args = 
to site.cfg under openblas header but that didn't seem to change anything. When it is building it looks like openblas_info:
/Users/lucasadams/numpy/numpy/distutils/system_info.py:633: UserWarning: Specified path  is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
  FOUND:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/opt/OpenBLAS/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/opt/OpenBLAS/lib']

  FOUND:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/opt/OpenBLAS/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/opt/OpenBLAS/lib']

and then a bit further down

  FOUND:
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
    extra_compile_args = ['-msse3']

**Comment by fgregg** [comment]
I don't know. maybe ask on stack overflow.

**Comment by evz** [comment]
@lucaswiser There's a PR to update the docs from @rud who noticed that the Homebrew formula is now referencing a newer version of OpenBLAS. Not sure if that helps your situation or makes it more complicated but it might be another angle of attack anyways.

**Comment by rud** [comment]
PR is this: https://github.com/datamade/dedupe/pull/429

**Comment by fgregg** [comment]
Closing for now. If you anyone wants to pick back up, let me know.

**Comment by gaving** [comment]
I'm clutching at straws here in this old thread, but I don't suppose anybody had any suggestions or tips on going about entity resolution using Spark?

**Comment by Alexkane** [comment]
I'm also interested in running Dedupe on Spark, has there been any recent activity on this?
